_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,791 @@
+            .await;
+
+        logs.push(WebsiteInfo {
+            url: website.url,
+            alias: website.alias,
+            data,
+        });
+    }
+
+    WebsiteLogs { logs }
+}
+```
+
+Once done, we’ll want to create a dynamic route for getting more information about a monitored URL. We can use this page to display things like past incidents/alerts. This function will follow most of the previous handler function except we’re fetching one URL and additionally grabbing any records of HTTP requests that didn’t return `200 OK` and labelling them as “Incidents”:
+
+```rust
+async fn get_website_by_id(
+    State(state): State<AppState>,
+    Path(alias): Path<String>,
+) -> impl AskamaIntoResponse {
+    let website = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites WHERE alias = $1")
+        .bind(&alias)
+        .fetch_one(&state.db)
+        .await
+        .unwrap();
+
+    // .. rest of your code
+}
+```
+
+Then we can write the previous code for getting the uptime percentages per hour, as well as grabbing a vector of `Incident`s:
+
+```rust
+// main.rs
+let last_24_hours_data = get_website_uptime_stats(SplitBy::Hour, &website.alias, &state.db).await;
+let monthly_data = get_website_uptime_stats(SplitBy::Day, &website.alias, &state.db).await;
+
+let incidents = sqlx::query_as::<_, Incident>(
+        "SELECT created_at as time, status from logs where website_alias = $1
+        and status != 200",
+    )
+    .bind(&alias)
+    .fetch_all(&state.db)
+    .await
+    .unwrap();
+
+    let log = WebsiteInfo {
+        url: website.url,
+        alias,
+        data,
+    };
+
+SingleWebsiteLogs { log, incidents, monthly_data }
+```
+
+“Incidents” in this case are defined as failed requests to a monitored URL. If the SQL query for getting Incidents fails (for example: because there aren’t any reported yet), Askama will automatically turn it into a vector of length zero instead of erroring out and causing a HTTP error.
+
+Here’s the full handler function if you get lost:
+
+```rust
+async fn get_website_by_alias(
+    State(state): State<AppState>,
+    Path(alias): Path<String>,
+) -> impl AskamaIntoResponse {
+    let website = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites WHERE alias = $1")
+        .bind(&alias)
+        .fetch_one(&state.db)
+        .await
+        .unwrap();
+
+    let last_24_hours_data = get_website_uptime_stats(SplitBy::Hour, &website.alias, &state.db).await;
+    let monthly_data = get_website_uptime_stats(SplitBy::Day, &website.alias, &state.db).await;
+
+    let incidents = sqlx::query_as::<_, Incident>(
+        "SELECT created_at as time, status from logs where website_alias = $1 and status != 200",
+    )
+    .bind(&alias)
+    .fetch_all(&state.db)
+    .await
+    .unwrap();
+
+    let log = WebsiteInfo {
+        url: website.url,
+        alias,
+        data: last_24_hours_data,
+    };
+
+    SingleWebsiteLogs { log, incidents, monthly_data }
```

Comment:
`data` got fixed here. But `SingleWebsiteLogs` still does not have a monthly field